pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,396 @@
+				scopedLog.Debug("Skipping migration of egress rule due to no routes found")
+				continue
+			}
+
+			oldTable := r.Table
+
+			// It is sufficient to grab the first route that matches because we
+			// are assuming all routes created under a rule will have the same
+			// ifindex (LinkIndex).
+			linkidx := routes[0].LinkIndex
+			newTable, err := retrieveTableIDFromIfIndex(linkidx, getter)
+			if err != nil {
+				overallErr = fmt.Errorf("failed to retrieve new table ID from link index %q: %w",
+					linkidx, err)
+				scopedLog.WithError(overallErr).
+					Error("Failed to retrieve new table ID during ENI datapath migration")
+				break
+			}
+
+			if revert, err := onUpgradeInstallRulesAndRoutes(
+				rules,
+				routes,
+				oldTable, newTable,
+			); err != nil {
+				overallErr = fmt.Errorf("failed to migrate ENI datapath: %w", err)
+				scopedLog.WithError(overallErr).Error("Failed to migrate ENI datapath")
+
+				if revErr := revertStack.Revert(); revErr != nil {
+					revertErr = fmt.Errorf("failed to revert ENI datapath migration: %w", err)
+					scopedLog.WithError(revertErr).Error("Failed to revert ENI datapath migration")
+				}
+
+				break
+			} else {
+				revertStack.Extend(revert)
+			}
+
+		case linux_defaults.RulePriorityEgressv2:
+			// Downgrade case. Unsupported for now.
+		}
+	}
+
+	if overallErr != nil {
+		return overallErr, revertErr
+	}
+
+	for _, r := range rules {
+		switch r.Priority {
+		case linux_defaults.RulePriorityEgress:
+			scopedLog := log.WithField("rule", r)
+
+			routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
+				Table: r.Table,
+			}, netlink.RT_FILTER_TABLE)
+			if err != nil {
+				overallErr = fmt.Errorf("failed to list rules routes: %w", err)
+				scopedLog.WithError(overallErr).
+					Debug("Skipping migration of egress rule due to failure in listing routes")
+				break
+			}
+
+			// If there are no routes under the same table as the rule, then
+			// skip.
+			if len(routes) == 0 {
+				scopedLog.Debug("Skipping migration of egress rule due to no routes found")
+				continue
+			}
+
+			oldTable := r.Table
+
+			if revert, err := onUpgradeDeleteRulesAndRoutes(
+				rules,
+				routes,
+				oldTable, linux_defaults.RulePriorityEgress,
+			); err != nil {
+				overallErr = fmt.Errorf("failed to remove old ENI datapath: %w", err)
+				scopedLog.WithError(overallErr).Error("Failed to migrate ENI datapath")
+
+				if revErr := revertStack.Revert(); revErr != nil {
+					revertErr = fmt.Errorf("failed to revert removal of old ENI datapath: %w", err)
+					scopedLog.WithError(revErr).Error("Failed to revert ENI datapath cleanup")
+				}
+
+				break
+			} else {
+				revertStack.Extend(revert)
+			}
+
+		case linux_defaults.RulePriorityEgressv2:
+			// Downgrade case. Unsupported for now.
+		}
+	}
+
+	if overallErr != nil {
+		if revErr := revertStack.Revert(); revErr != nil {
+			log.WithError(revErr).Error("Failed to revert ENI full datapath migration")
+		}
```

Comment:
Seems like the revertstack will be run twice in this case due to lines 152-155, which is highly likely to cause any failure in the second portion to result in error logs here.

Reply:
Fixed.